File: Backtracking/46_Permutations/solution46.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        
        if (nums.size() == 1) {
            return {nums};
        }
        
        for (int i = 0; i < nums.size(); ++i) {
            int n = nums[0];
            nums.erase(nums.begin());
            
            vector<vector<int>> perms = permute(nums);
            
            for (auto& perm : perms) {
                perm.push_back(n);
            }
            
            res.insert(res.end(), perms.begin(), perms.end());
            
            nums.push_back(n);
        }
        
        return res;
    }
};

// Runtime1: 3ms, beats 52.66%
// Memory1: 11.91MB, beats 6.79%

// Runtime2: 3ms, beats 52.66%
// Memory2: 12.12MB, beats 6.40%

// Runtime3: 0ms, beats 100%
// Memory3: 12.11MB, beats 6.40%

File: Backtracking/46_Permutations/solution46.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []

        if len(nums) == 1:
            return [nums[:]]

        for i in range(len(nums)):
            n = nums.pop(0)
            perms = self.permute(nums)

            for perm in perms:
                perm.append(n)
            res.extend(perms)
            nums.append(n)
        return res

# Runtime1: 39ms, beats 66.97%
# Memory1: 16.80MB, beats 34.53%

# Runtime2: 39ms, beats 66.97% 
# Memory2: 16.61MB, beats 70.90%

# Runtime3: 39ms, beats 66.97% 
# Memory3: 16.75MB, beats 34.53%